### PR TITLE
Serialize F# DUs in a partial trust environment

### LINKED ISSUE #821: Exception in Partial Trust Environment
I'm attempting to serialize F# union types in a Silverlight app. When I attempt this, I get an exception:

```
System.MethodAccessException occurred
  Message=Attempt by method 'System.Linq.Expressions.Expression.CreateLambda(System.Type, System.Linq.Expressions.Expression, System.String, Boolean, System.Collections.ObjectModel.ReadOnlyCollection`1<System.Linq.Expressions.ParameterExpression>)' to access method 'System.Linq.Expressions.Expression`1<Newtonsoft.Json.Utilities.MethodCall`2<System.Object,System.Object>>.Create(System.Linq.Expressions.Expression, System.String, Boolean, System.Collections.ObjectModel.ReadOnlyCollection`1<System.Linq.Expressions.ParameterExpression>)' failed.
  StackTrace:
       at System.RuntimeMethodHandle.PerformSecurityCheck(Object obj, RuntimeMethodHandleInternal method, RuntimeType parent, UInt32 invocationFlags)
       at System.RuntimeMethodHandle.PerformSecurityCheck(Object obj, IRuntimeMethodInfo method, RuntimeType parent, UInt32 invocationFlags)
       at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture, Boolean skipVisibilityChecks)
       at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
       at System.Linq.Expressions.Expression.CreateLambda(Type delegateType, Expression body, String name, Boolean tailCall, ReadOnlyCollection`1 parameters)
       at System.Linq.Expressions.Expression.Lambda(Type delegateType, Expression body, String name, Boolean tailCall, IEnumerable`1 parameters)
       at Newtonsoft.Json.Utilities.ExpressionReflectionDelegateFactory.CreateMethodCall[T](MethodBase method)
       at Newtonsoft.Json.Utilities.FSharpUtils.EnsureInitialized(Assembly fsharpCoreAssembly)
       at Newtonsoft.Json.Converters.DiscriminatedUnionConverter.CanConvert(Type objectType)
       at Newtonsoft.Json.JsonSerializer.GetMatchingConverter(IList`1 converters, Type objectType)
       at Newtonsoft.Json.Serialization.DefaultContractResolver.InitializeContract(JsonContract contract)
       at Newtonsoft.Json.Serialization.DefaultContractResolver.CreateObjectContract(Type objectType)
       at Newtonsoft.Json.Serialization.DefaultContractResolver.CreateContract(Type objectType)
       at Newtonsoft.Json.Serialization.DefaultContractResolver.ResolveContract(Type type)
       at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.CalculatePropertyValues(JsonWriter writer, Object value, JsonContainerContract contract, JsonProperty member, JsonProperty property, JsonContract& memberContract, Object& memberValue)
       at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeObject(JsonWriter writer, Object value, JsonObjectContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
       at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeValue(JsonWriter writer, Object value, JsonContract valueContract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerProperty)
       at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeList(JsonWriter writer, IEnumerable values, JsonArrayContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
       at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeValue(JsonWriter writer, Object value, JsonContract valueContract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerProperty)
       at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.Serialize(JsonWriter jsonWriter, Object value, Type objectType)
       at Newtonsoft.Json.JsonSerializer.SerializeInternal(JsonWriter jsonWriter, Object value, Type objectType)
       at Newtonsoft.Json.JsonSerializer.Serialize(JsonWriter jsonWriter, Object value)
```

From some brief code-surfing, I suspect the issue may be that the `BindingFlags` [here](https://github.com/JamesNK/Newtonsoft.Json/blob/master/Src/Newtonsoft.Json/Utilities/FSharpUtils.cs#L91) include `BindingFlags.NonPublic`, which is a no-no in partial trust.

### PR DESCRIPTION
**Pull request by floyd-may**
This fixes #821. I had to decorate the tests assembly to allow partial trust callers; the same will also need to be true for any F# assembly that has DUs that need to be serialized.

### CODE REVIEW

**Comment by floyd-may** [comment]
Apologies for the build churn :( I'll have it in a few

**Comment by floyd-may** [comment]
Okay, I could use a hand here :cold_sweat:

File: Src/Newtonsoft.Json.Tests/Properties/AssemblyInfo.cs [hunk 1]
```diff
@@ -43,11 +45,14 @@
 [assembly: AssemblyTitle("Json.NET Tests .NET 3.5")]
 #elif NET40
 [assembly: AssemblyTitle("Json.NET Tests .NET 4.0")]
+[assembly: SecurityRules(SecurityRuleSet.Level1)]
 #else
-
 [assembly: AssemblyTitle("Json.NET Tests")]
+[assembly: SecurityRules(SecurityRuleSet.Level1)]
```
Comment (floyd-may) [inline Src/Newtonsoft.Json.Tests/Properties/AssemblyInfo.cs]: This is the LOC that's causing the build issues. If it's present, we get the `System.IO.FileLoadException` complaining about the `SecurityRules` attribute; if I get rid of it, I get security transparency rule failures, but only under the DNXCore 5.0 test run. Not quite sure where to go from here.

**Comment by JamesNK** [comment]
I believe I have fixed it here - https://github.com/JamesNK/Newtonsoft.Json/commit/eacb191f97cf460934d61d0a62ae8d3680c70528

**Pull request closed without merging**